Add a maximum path limit for each cycle to the HawickJames algorithm

EdwinOuwehand opened the pull request:
Motivation: In its current form the algorithm does not perform well when many cycles exist. For example, 40 vertices, 400 edges and a high number of cycles runs out of memory with 16G. In this case it makes sense to limit search depth, because the majority of the longer paths are probably already part of a smaller cycle and therefore don't have to be explored.

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

EdwinOuwehand commented:
Motivation: In its current form the algorithm does not perform well when many cycles exist. For example, 40 vertices, 400 edges and a high number of cycles runs out of memory with 16G. In this case it makes sense to limit search depth, because the majority of the longer paths are probably already part of a smaller cycle and therefore don't have to be explored.

jsichi commented:
Thanks for adding the unit test class...seems like we're missing a few more from way back when we adopted the cycles package.

But could you avoid reformatting the existing code?  It makes it harder to review your changes.

EdwinOuwehand commented:
> But could you avoid reformatting the existing code? It makes it harder to review your changes.

@jsichi I just removed some duplicate code, simplified the count/find/print logic and then ran the auto-formatter. For actually limiting the paths all I did was add the getter/setter and the limitReached check. 

Would you like me to revert some of the changes, or leave it as is?

jsichi commented:
When using the autoformatter, be sure to import our code style settings as described in

https://github.com/jgrapht/jgrapht/wiki/How-to-setup-your-development-environment-for-JGraphT

That will minimize spurious updates.

But it's simpler to just skip autoformatting, since we run it in bulk as part of each release.

Could you revert just the autoformat changes?

EdwinOuwehand commented:
Sure, done ;)

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HawickJamesSimpleCycles.java:
```diff
@@ -213,36 +188,15 @@ private void unblock(Integer u)
             Integer w = B[u].get(wPos);
 
-            wPos -= removeFromList(B[u], w);
+            wPos -= frequency(B[u], w);
```
The old removeFromList code does the counting and removal in a single pass.  Your new code requires two passes.  I think you can eliminate the frequency call by subtracting the B[u].size before and after the removeAll.

maxmousee replied:
Done, thanks:)

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HawickJamesSimpleCycles.java:
```diff
@@ -327,20 +268,49 @@ public long countSimpleCycles()
+     * This is the default behaviour of the algorithm.
+     * It will keep looking as long as there are paths available.
+     */
+    public void unlimitedPaths()
```
Rename this method to clearPathLimit

maxmousee replied:
Done:)

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HawickJamesSimpleCycles.java:
```diff
@@ -213,36 +188,16 @@ private void unblock(Integer u)
-            wPos -= removeFromList(B[u], w);
+            wPos -= B[u].size();
+            B[u].removeAll(singletonList(w));
+            wPos -= B[u].size();
```
Errr, I meant:

```
            int sizeBeforeRemove = B[u].size();
            B[u].removeAll(singletonList(w));
            wPos -= (sizeBeforeRemove - B[u].size());
```

maxmousee replied:
Oh, sorry, thanks for the clarification:) 
I fixed it:)

jsichi commented:
Thanks!

dennis-sig commented:
@jsichi would it be possible to make a minor release that includes this fix? This issue prevents us from using JGraphT because we work with some very large graphs.

d-michail commented:
We automatically build snapshot releases on each merge into master. See [using-via-maven](https://github.com/jgrapht/jgrapht/blob/master/README.md#using-via-maven) for the necessary dependency using maven.
